File: asc/dashboards/page/facilities_map/facilities_map.py
import frappe
import json

@frappe.whitelist()
def get_data(basic_facility = None, year = None, division = None):
    facility = basic_facility
    case_string=""
    division_case_string=""
    if division:
        case_string=" and region = '%s' " %(str(division))
        division_case_string= " and k.division = '%s' " %(str(division))
    facility_name = get_facility(facility,year)

    temp_query = """Select k.district as name,
     %s , 
    d.path_features as path
    from tabASC_KPI k
    Left join 
    tabDistrict d 
    on 
    k.district = d.name where k.district != 'Keamari Karachi' and Year = '%s' %s group by k.district"""%(facility_name,str(year),division_case_string)
    # frappe.msgprint(frappe.as_json(temp_query))
    
    data = frappe.db.sql(temp_query, as_dict=1) 
#     frappe.msgprint(frappe.as_json(data))

    data = sorted(data, key=lambda x: x['value'], reverse=True)
    # frappe.msgprint(frappe.as_json((basic_facility.replace(" ","_")).lower()))

    color_query = """Select `min` as 'from', (CASE WHEN `max`= 0 then '' else `max` end) as 'to' , `color` as 'color' , `range` as 'name' from `tabColor Indication` where parentfield = '%s' order by `order` asc"""%((basic_facility.replace(" ","_")).lower())
    color_data =  frappe.db.sql(color_query, as_dict=1)  
    for row in color_data:
        if row['to'] == '':
            del row['to']
    data_ =[]
    data_.append(data) 
    data_.append(color_data)

    return data_



def get_facility(facility,year):
    query = ""
    if facility == "Toilet":
        query = "COUNT(k.name) AS Total_Schools , IFNULL(SUM(CASE WHEN toilet ='Yes' THEN 1 ELSE 0 END),0)  as facility_available, ROUND(IFNULL(SUM(CASE WHEN toilet ='Yes' THEN 1 ELSE 0 END),0) / count(k.name) *100 ,1) as value"
    elif facility == "Drinking Water":
        query = "COUNT(k.name) AS Total_Schools, IFNULL(SUM(if(water = 'Yes'  , 1, 0)),0) as facility_available ,  ROUND(IFNULL(SUM(if(water = 'Yes'  , 1, 0)),0) / count(k.name) *100 ,1) as value"
    elif facility == "Electricity":
        query = "COUNT(k.name) AS Total_Schools, IFNULL(SUM(if(electricity = 'No Electricity Connection' OR electricity='' OR electricity IS NULL,0,1)),0) as facility_available ,ROUND(IFNULL(SUM(if(electricity = 'No Electricity Connection' OR electricity='' OR electricity IS NULL,0,1)),0) / count(k.name) *100 ,1)as value  "
    elif facility == "Building":
        query = "COUNT(k.name) AS Total_Schools, IFNULL(SUM(CASE WHEN `building_availability`='Yes' THEN 1 ELSE 0 END),0) as facility_available, ROUND(IFNULL(SUM(CASE WHEN `building_availability`='Yes' THEN 1 ELSE 0 END),0) / count(k.name) *100 ,1)as value "
    elif facility == "Boundary Wall":
        query = "COUNT(k.name) AS Total_Schools, IFNULL(SUM(if(`condition_of_boundary_wall`='No Boundary Wall' OR `condition_of_boundary_wall`='' OR `condition_of_boundary_wall` IS NULL, 0, 1) ),0) as facility_available, ROUND(IFNULL(SUM(if(`condition_of_boundary_wall`='No Boundary Wall' OR `condition_of_boundary_wall`='' OR `condition_of_boundary_wall` IS NULL, 0, 1) ),0) / count(k.name) *100 ,1)as value "
    elif facility == "Hand Wash":
        query = "COUNT(k.name) AS Total_Schools, IFNULL(SUM(if(`hand_wash`='Yes', 1, 0) ),0) as facility_available, ROUND(IFNULL(SUM(if(`hand_wash`='Yes', 1, 0) ),0)  / count(k.name) *100,1) as value   "
    elif facility == "Soap":
        query = "COUNT(k.name) AS Total_Schools, IFNULL(SUM(if(`soap`='Yes', 1, 0) ),0) as facility_available, ROUND(IFNULL(SUM(if(`soap`='Yes', 1, 0) ),0)  / count(k.name) *100 ,1) as value   "
    elif facility == "MHM Disposal":
        query = "SUM(CASE WHEN gender != 'Boys' and level != 'Primary' and Year = '%s' then 1 else 0 end) AS Total_Schools, IFNULL(SUM(if(`mhm_facility`='Yes', 1, 0) ),0) as facility_available, ROUND(IFNULL(SUM(if(`mhm_facility`='Yes', 1, 0) ),0)  / SUM(CASE WHEN gender != 'Boys' and level != 'Primary' and Year = '%s' then 1 else 0 end) *100 ,1) as value   "%(year,year)
    return query


@frappe.whitelist()
def get_taluka_data(basic_facility = None, year = None, district = None):

    facility = basic_facility
    case_string=""
    division_case_string=""
    if district:
        division_case_string= " and k.district = '%s' " %(str(district))
    facility_name = get_facility(facility,year)


    temp_query = """Select k.district as name,
     Count(k.name) as total_schools , 
     Sum(Case when status = 'Functional' then 1 else 0 end ) as funcional,
     Sum(Case when status = 'Closed' then 1 else 0 end ) as closed,

     Sum(Case when gender = 'Boys' then 1 else 0 end ) as boys,
     Sum(Case when gender = 'Girls' then 1 else 0 end ) as girls,
     Sum(Case when gender = 'Mixed' then 1 else 0 end ) as mixed,

     Sum(Case when level = 'Primary' then 1 else 0 end ) as 'primary',
     Sum(Case when level = 'Elementary' then 1 else 0 end ) as elementry,
     Sum(Case when level = 'Middle' then 1 else 0 end ) as middle,
     Sum(Case when level = 'Secondary' then 1 else 0 end ) as secondary,
     Sum(Case when level = 'Higher Secondary' then 1 else 0 end ) as higher_secondary,

     SUM( IFNULL(`boys_enrollment`,0)) as male_enrollment,
     SUM( IFNULL(`girls_enrollment`,0)) as female_enrollment,
     SUM(IFNULL(`male_teachers`,0)) as male_teachers,
     SUM(IFNULL(`female_teachers`,0)) as female_teachers,

     SUM(if(water = 'Yes'  , 1, 0)) as water,
     SUM(if(`condition_of_boundary_wall`='No Boundary Wall' OR `condition_of_boundary_wall`='' OR `condition_of_boundary_wall` IS NULL, 0, 1) ) as boundary_wall,
     SUM(if(`hand_wash`='Yes', 1, 0) ) as hand_wash,
     SUM(if(`mhm_facility`='Yes', 1, 0) ) as mhm,
     SUM(CASE WHEN gender != 'Boys' and level != 'Primary'  then 1 else 0 end) AS total_mhm_Schools,
     SUM(CASE WHEN `building_availability`='Yes' THEN 1 ELSE 0 END) as building,
     SUM(CASE WHEN toilet ='Yes' THEN 1 ELSE 0 END) as toilet,
     SUM(if(electricity = 'No Electricity Connection' OR electricity='' OR electricity IS NULL,0,1)) as electricity,
     SUM(if(`soap`='Yes', 1, 0) ) as soap

    from tabASC_KPI k
     where Year = '%s'  %s """%(str(year),division_case_string)
    # frappe.msgprint(frappe.as_json(temp_query))
    
    data_ = frappe.db.sql(temp_query, as_dict=1) 

    if data_[0]['total_schools']:
        if data_[0]['primary']:
            data_[0]['primary_percentage'] = round(data_[0]['primary'] / data_[0]['total_schools'] * 100,1)
        else:
            data_[0]['primary_percentage'] = 0
        if data_[0]['elementry']:
                data_[0]['elementry_percentage'] = round(data_[0]['elementry'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['elementry_percentage'] = 0
        if data_[0]['middle']:
                data_[0]['middle_percentage'] = round(data_[0]['middle'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['middle_percentage'] = 0       
        if data_[0]['secondary']:
                data_[0]['secondary_percentage'] = round(data_[0]['secondary'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['secondary_percentage'] = 0 
        if data_[0]['higher_secondary']:
                data_[0]['higher_secondary_percentage'] = round(data_[0]['higher_secondary'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['higher_secondary_percentage'] = 0 
        if data_[0]['water']:
                data_[0]['water_percentage'] = round(data_[0]['water'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['water_percentage'] = 0 
        if data_[0]['boundary_wall']:
                data_[0]['boundary_wall_percentage'] = round(data_[0]['boundary_wall'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['boundary_wall_percentage'] = 0 
        if data_[0]['hand_wash']:
                data_[0]['hand_wash_percentage'] = round(data_[0]['hand_wash'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['hand_wash_percentage'] = 0 
        if data_[0]['mhm'] and data_[0]['total_mhm_Schools']:
                data_[0]['mhm_percentage'] = round(data_[0]['mhm'] / data_[0]['total_mhm_Schools'] * 100,1)
        else:
                data_[0]['mhm_percentage'] = 0 
        if data_[0]['building'] :
                data_[0]['building_percentage'] = round(data_[0]['building'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['building_percentage'] = 0 
        if data_[0]['toilet'] :
                data_[0]['toilet_percentage'] = round(data_[0]['toilet'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['toilet_percentage'] = 0 
        if data_[0]['electricity'] :
                data_[0]['electricity_percentage'] = round(data_[0]['electricity'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['electricity_percentage'] = 0
        if data_[0]['soap'] :
                data_[0]['soap_percentage'] = round(data_[0]['soap'] / data_[0]['total_schools'] * 100,1)
        else:
                data_[0]['soap_percentage'] = 0
    # return_data={'District': data}

    temp_query_ = """Select k.tehsil as name,
     %s , 
    d.path_features as path
    from tabASC_KPI k
    Left join 
    tabDistrict d 
    on 
    k.district = d.name where Year = '%s' %s group by k.tehsil"""%(facility_name,str(year),division_case_string)
    # frappe.msgprint(frappe.as_json(temp_query_))
    
    data = frappe.db.sql(temp_query_, as_dict=1) 
    data = sorted(data, key=lambda x: x['value'], reverse=True)
    return_data={'District': data_,
                'Taluka': data}

    return return_data

File: asc/dashboards/page/facilities_map/facilities_map.js
frappe.pages['facilities-map'].on_page_load = function (wrapper) {
	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'Facilities Map',
		single_column: true,

	});
	console.log("Route Data", frappe.get_route());
	$(".select-icon").css('top: 7px');

	filters.add(page);
	// $("input[data-fieldname='basic_facility']").css('align-items: center;');
	$(".select-icon").css('top: 7px');

	// $(".frappe-control[data-fieldtype='Select'], .form-group, .select-icon.xs").css('top: 7px');
	var basic_facility = "Drinking Water"
	var year = "2021-22"

	if (frappe.get_route().length > 1) {
		if (frappe.get_route()[1]) {
			year = frappe.get_route()[1]
		}
		if (frappe.get_route()[2]) {
			basic_facility = frappe.get_route()[2]
		}
	}

	frappe.call({
		method: "asc.dashboards.page.facilities_map.facilities_map.get_data",
		freeze: true,
		args: {
			basic_facility: basic_facility,
			year: year,
			freeze: true,
		},

		callback: function (r) {
			// console.log("Callback");
			//console.log(r.message);
			// console.log("Callback end");
			var data = r.message[0];

			$('#mainBody').remove();


			$(
				frappe.render_template("facilities_map", { 'type': basic_facility })
			).appendTo(page.main);
			$("#profile_items").hide()
			$("#barContainer").hide()


			draw_map(r.message[0], r.message[1], basic_facility, year, page);
			legend = true


			barChart2(r.message[0], basic_facility, 'barContainer2', legend, 'bar')

			//provinceChart2(data, r.message[1], basic_facility.get_value());
		},
	});

}

filters = {
	add: function (page) {
		var basic_facility_ = "Drinking Water"
		var year_ = "2021-22"

		if (frappe.get_route().length > 1) {
			if (frappe.get_route()[1]) {
				year_ = frappe.get_route()[1]
			}
			if (frappe.get_route()[2]) {
				basic_facility_ = frappe.get_route()[2]
			}
		}
		let year = page.add_field({
			label: "Year",
			fieldtype: "Link",
			fieldname: "Year",
			options: "Year",
			default: year_,
			reqd: 1,
		});
		let division = page.add_field({
			label: "Division",
			fieldtype: "Link",
			fieldname: "division",
			options: "Division",
		});
		let basic_facility = page.add_field({
			label: "Basic Facility",
			fieldtype: "Select",
			fieldname: "basic_facility",
			default: basic_facility_,
			options: ["", "Drinking Water", "Electricity", "Building", "Boundary Wall", "Toilet", "Hand Wash", "Soap", "MHM Disposal"],
			reqd: 1,
		});
		let fiterbtn = page.add_field({
			label: "View",
			fieldtype: "Button",
			fieldname: "filter",
			click() {
				if (basic_facility.get_value() == "" || year.get_value() == "") {
					if (year.get_value() == "" && basic_facility.get_value() == "") {
						frappe.msgprint("Please Select Year and Facility");
						return;
					}
					if (year.get_value() == "") {
						frappe.msgprint("Please Select Year");
						return;
					}
					if (basic_facility.get_value() == "") {
						frappe.msgprint("Please Select Facility");
						return;
					}

				} else {
					//console.log("Inside Click");


					frappe.call({
						method: "asc.dashboards.page.facilities_map.facilities_map.get_data",
						freeze: true,
						args: {
							basic_facility: basic_facility.get_value(),
							year: year.get_value(),
							division: division.get_value(),
							freeze: true,
						},

						callback: function (r) {
							// console.log("Callback");
							//console.log(r.message);
							// console.log("Callback end");
							var data = r.message[0];

							$('#mainBody').remove();


							$(
								frappe.render_template("facilities_map", { 'type': basic_facility.get_value() })
							).appendTo(page.main);
							$("#profile_items").hide()
							$("#barContainer").hide()


							draw_map(r.message[0], r.message[1], basic_facility.get_value(), year.get_value(), page);
							legend = true
							if (division.get_value() != "") {
								legend = true
							}

							barChart2(r.message[0], basic_facility.get_value(), 'barContainer2', legend, 'bar')

							//provinceChart2(data, r.message[1], basic_facility.get_value());
						},
					});
				}
			},
		});






	},
};//end of filters

function barChart(data_, facility_name, container, legend, type, height) {
	categoery = []
	facility_available = []
	facility_not_available = []
	for (let i = 0; i < data_.length; i++) {
		categoery[i] = data_[i].name;
		facility_available[i] = data_[i].value;
		facility_not_available[i] = parseFloat((100 - data_[i].value).toFixed(1));
	}
	console.log(facility_available);
	console.log(facility_not_available);

	Highcharts.chart(container, {
		chart: {
			type: type,
			height: 350,
			// width: 450,
		},
		credits: {
			enabled: false,
		},
		title: {
			text: facility_name + ' Facility',
			align: 'left',
			style: {
				fontWeight: 'bold',
				fontSize: '16px',
			}
		},
		xAxis: {
			categories: categoery,
			crosshair: true
		},
		yAxis: {
			min: 0,
			title: {
				text: ""
			}
		},
		legend: {
			reversed: true,
			enabled: legend,
		},

		plotOptions: {
			series: {
				pointWidth: 28,
				stacking: 'percent',
				dataLabels: {
					enabled: true,
					formatter: function () {
						if (this.y) {
							return this.y;
						}
					},
					position: "left",
					allowOverlap: true,
					crop: false,
					padding: 0,

				},
			}
		},
		series: [{
			name: 'Schools with ' + facility_name,
			data: facility_available,
			color: '#345C0C',
		}, {
			name: 'Other Schools',
			data: facility_not_available,
			color: '#EAB10B',


		}]
	});


}
function barChart2(data_, facility_name, container, legend, type, height) {
	categoery = []
	facility_available = []
	facility_not_available = []
	for (let i = 0; i < data_.length; i++) {
		categoery[i] = data_[i].name;
		facility_available[i] = data_[i].value;
		facility_not_available[i] = parseFloat((100 - data_[i].value).toFixed(1));
	}
	console.log(facility_available);
	console.log(facility_not_available);

	Highcharts.chart(container, {
		chart: {
			type: type,
			// width: 450,
		},
		credits: {
			enabled: false,
		},
		title: {
			text: facility_name + ' Facility',
			align: 'left',
			style: {
				fontWeight: 'bold',
				fontSize: '16px',
			}
		},
		xAxis: {
			categories: categoery,
			crosshair: true
		},
		yAxis: {
			min: 0,
			title: {
				text: ""
			}
		},
		legend: {
			reversed: true,
			enabled: legend,
		},

		plotOptions: {
			series: {
				pointWidth: 28,
				stacking: 'percent',
				dataLabels: {
					enabled: true,
					formatter: function () {
						if (this.y) {
							return this.y;
						}
					},
					position: "left",
					allowOverlap: true,
					crop: false,
					padding: 0,

				},
			}
		},
		series: [{
			name: 'Schools with ' + facility_name,
			data: facility_available,
			color: '#345C0C',
		}, {
			name: 'Other Schools',
			data: facility_not_available,
			color: '#EAB10B',


		}]
	});


}




function draw_map(adata, color, facility, year, page) {
	var series = [{
		type: 'map',
		enableMouseTracking: true,
		showInLegend: false,
		animation: {
			duration: 1000
		},
		data: adata,
		dataLabels: {
			enabled: true,
			color: '#FFFFFF',
			format: '{point.name} <br />( {point.value} %)'
		},
		name: facility + ' Availability',
		states: {
			hover: {
				borderColor: '#FFFFFF'
			}
		},
		tooltip: {
			pointFormat: 'District: {point.name} <br /> Total Schools: {point.Total_Schools}<br />Schools with ' + facility + ': {point.facility_available} ({point.value}%)<br /><br />'
		}
	}];
	Highcharts.mapChart('mapContainer', {
		title: {
			text: ''
		},
		mapNavigation: {
			enabled: true,
			buttonOptions: {
				alignTo: 'spacingBox',
				x: 10
			}
		},
		colorAxis: {
			dataClasses: color
		},
		legend: {
			layout: 'vertical',
			align: 'left',
			verticalAlign: 'bottom',
			floating: true,

		},
		plotOptions: {
			series: {
				events: {
					click: function (e) {
						frappe.call({
							method: "asc.dashboards.page.facilities_map.facilities_map.get_taluka_data",
							freeze: true,
							args: {
								basic_facility: facility,
								year: year,
								district: e.point.name,
								freeze: true,
							},

							callback: function (r) {
								console.log("Callback2");
								console.log(r.message);
								console.log("Callback 2 end");

								// $('#mainBody').remove();


								// $(
								// 	frappe.render_template("facilities_map", { 'type': facility })
								// ).appendTo(page.main);

								// draw_map(r.message.District, facility, year, page);
								$("#profile_items").show()
								$("#barContainer").show()
								$("#barContainer2").hide()
								$('[data-toggle="tooltip"]').tooltip();
								// console.log(r.message.District[0].name);
								$('#district_name').html(r.message.District[0].name)

								$('#school').html((r.message.District[0].total_schools).toLocaleString("en-US"))

								$('#functional').html((r.message.District[0].funcional).toLocaleString("en-US"))
								$('#closed').html((r.message.District[0].closed).toLocaleString("en-US"))

								$('#male_enrollment').html((r.message.District[0].male_enrollment).toLocaleString("en-US"))
								$('#female_enrollment').html((r.message.District[0].female_enrollment).toLocaleString("en-US"))
								$('#enrollment').html((r.message.District[0].male_enrollment + r.message.District[0].female_enrollment).toLocaleString("en-US"))


								$('#teachers').html((r.message.District[0].male_teachers + r.message.District[0].female_teachers).toLocaleString("en-US"))
								$('#male_teachers').html((r.message.District[0].male_teachers).toLocaleString("en-US"))
								$('#female_teachers').html((r.message.District[0].female_teachers).toLocaleString("en-US"))

								var data = String(r.message.District[0].primary.toLocaleString("en-US")) + " (" + String(r.message.District[0].primary_percentage) + "%)"
								$('#primary').html(data)

								var data = String(r.message.District[0].elementry.toLocaleString("en-US")) + " (" + String(r.message.District[0].elementry_percentage) + "%)"
								$('#elementary').html(data)

								var data = String(r.message.District[0].middle.toLocaleString("en-US")) + " (" + String(r.message.District[0].middle_percentage) + "%)"
								$('#middle').html(data)

								var data = String(r.message.District[0].secondary.toLocaleString("en-US")) + " (" + String(r.message.District[0].secondary_percentage) + "%)"
								$('#secondary').html(data)

								var data = String(r.message.District[0].higher_secondary.toLocaleString("en-US")) + " (" + String(r.message.District[0].higher_secondary_percentage) + "%)"
								$('#higher_secondary').html(data)


								$('#boys').html(r.message.District[0].boys.toLocaleString("en-US"))
								$('#girls').html(r.message.District[0].girls.toLocaleString("en-US"))
								$('#mixed').html(r.message.District[0].mixed.toLocaleString("en-US"))

								var data = String(r.message.District[0].hand_wash_percentage) + "%"
								$('#hand_wash_percentage').html(data);

								var data = String(r.message.District[0].boundary_wall_percentage) + "%"
								$('#boundary_wall_percentage').html(data)

								var data = String(r.message.District[0].water_percentage) + "%"
								$('#water_percentage').html(data);

								var data = String(r.message.District[0].mhm_percentage) + "%"
								$('#mhm_percentage').html(data)


								var data = String(r.message.District[0].building_percentage) + "%"
								$('#building_percentage').html(data)

								var data = String(r.message.District[0].toilet_percentage) + "%"
								$('#toilet_percentage').html(data)

								var data = String(r.message.District[0].electricity_percentage) + "%"
								$('#electricity_percentage').html(data)

								var data = String(r.message.District[0].soap_percentage) + "%"
								$('#soap_percentage').html(data)

								barChart(r.message.Taluka, facility, 'barContainer', false, 'column')
								$('#close_btn').click(function () {
									$("#profile_items").hide()
									$("#barContainer").hide()
									$("#barContainer2").show()
								});


							},
						});
					}
				}
			}
		},
		credits: {
			enabled: false,
		},
		exporting: {
			enabled: true
		},
		series: series
	});
}

